osh/cmd_parse.py

```diff
@@ -1422,31 +1427,38 @@ def ParseCase(self):
     case_node.to_match = to_match
     self._Next()
 
+    in_kw = None  # type: Token
+
     self._NewlineOk()
-    in_spid = word_.LeftMostSpanForWord(self.cur_word)
     self._Peek()
     if self.parse_opts.parse_brace() and self.c_id == Id.Lit_LBrace:
       self._Next()
     else:
       self._Eat(Id.KW_In)
+      in_kw = word_.AsKeywordToken(self.cur_word)
     self._NewlineOk()
 
     if self.c_id != Id.KW_Esac:  # empty case list
-      self.ParseCaseList(case_node.arms)
+      esac_kw = self.ParseCaseList(case_node.arms)
       # TODO: should it return a list of nodes, and extend?
       self._Peek()
+    else:
+        esac_kw = word_.AsKeywordToken(self.cur_word)
 
-    esac_spid = word_.LeftMostSpanForWord(self.cur_word)
     self._Peek()
     if self.parse_opts.parse_brace() and self.c_id == Id.Lit_RBrace:
       self._Next()
     else:
       self._Eat(Id.KW_Esac)
     self._Next()
 
-    case_node.spids.append(case_spid)
-    case_node.spids.append(in_spid)
-    case_node.spids.append(esac_spid)
+    case_node.case_kw = case_kw
+    case_node.in_kw = in_kw
+    case_node.esac_kw = esac_kw
+
+    case_node.spids.append(case_kw.span_id)
+    case_node.spids.append(in_kw.span_id if in_kw else runtime.NO_SPID)
+    case_node.spids.append(esac_kw.span_id if in_kw else runtime.NO_SPID)
```

Comment:
looks like a typo -- `esac_kw` ?

I'm surprised that didn't cause any tests to fail?  It looks like `test/ysh-prettify` is doing something with `esac`

Reply:
shouldn't it be `esac_kw.span_id if esac_kw else ...` ?

Reply:
Ah, sorry. I've fixed this.

The reason why the tests always passed is that `in_kw` and `esac_kw` are either both always defined or both always `None`. So technically the above code is correct, despite it being deceivingly so.

I've pushed a fix anyways.